driver: spi: silabs_siwx91x_gspi: Add pm device support for gspi driver

fimohame opened the pull request:
This commit enables the pm device driver support for the spi_silabs_siwx91x_gspi driver.

smalae commented on drivers/spi/spi_silabs_siwx91x_gspi.c:
```diff
@@ -58,6 +59,38 @@ struct gspi_siwx91x_data {
+			return ret;
+		}
+
+		cfg->reg->GSPI_CONFIG1_b.GSPI_MANUAL_CSN = 0;
```
It will affect functionality when operating at high speed

fimohame committed:
driver: spi: silabs_siwx91x_gspi: Add pm device support for gspi driver

This commit enables the pm device driver support
for the spi_silabs_siwx91x_gspi driver.

Signed-off-by: S Mohamed Fiaz <fiaz.mohamed@silabs.com>

jerome-pouiller commented on drivers/spi/spi_silabs_siwx91x_gspi.c:
```diff
@@ -617,7 +652,7 @@ static int gspi_siwx91x_init(const struct device *dev)
 	cfg->reg->GSPI_CONFIG1_b.GSPI_MANUAL_CSN = 0;
 
-	return 0;
+	return pm_device_driver_init(dev, gspi_siwx91x_pm_action);
```
I believe you can remove redundant parts between `gspi_siwx91x_pm_action()` and `gspi_siwx91x_init()`, no?

fimohame replied:
It's required to have them

fimohame replied:
I believe it's required

Martinhoff-maker commented on drivers/spi/spi_silabs_siwx91x_gspi.c:
```diff
@@ -591,6 +592,35 @@ static int gspi_siwx91x_release(const struct device *dev, const struct spi_confi
+		cfg->reg->GSPI_CONFIG1_b.GSPI_MANUAL_CSN = 0;
+		data->ctx.config = NULL;
+	} else if (IS_ENABLED(CONFIG_PM_DEVICE) && (action == PM_DEVICE_ACTION_SUSPEND)) {
+		return 0;
```
So nothing needs to be done is we go in sleep mode?

fimohame replied:
addressed

Martinhoff-maker commented on drivers/spi/spi_silabs_siwx91x_gspi.c:
```diff
@@ -199,7 +201,7 @@ static void gspi_siwx91x_dma_rx_callback(const struct device *dev, void *user_da
 
 	if (status >= 0 && status != DMA_STATUS_COMPLETE) {
-		return;
+		goto pm_put_exit;
```
you don't want to put the device runtime if the DMA transfer is not completed. in that case, you could turn off the power domain and then killed the working uDMA.

Martinhoff-maker commented on drivers/spi/spi_silabs_siwx91x_gspi.c:
```diff
@@ -533,8 +538,14 @@ static int gspi_siwx91x_transceive(const struct device *dev, const struct spi_co
+
 	if (!spi_siwx91x_is_dma_enabled_instance(dev) && asynchronous) {
 		ret = -ENOTSUP;
+		goto pm_put_exit;
```
you will never call the "pm_device_runtime_put" since you are in the case of an asynchronous transfer.

fimohame replied:
addressed

Martinhoff-maker commented on drivers/spi/spi_silabs_siwx91x_gspi.c:
```diff
@@ -543,7 +554,7 @@ static int gspi_siwx91x_transceive(const struct device *dev, const struct spi_co
 		if (ret) {
 			spi_context_release(&data->ctx, ret);
-			return ret;
+			goto pm_put_exit;
```
same here in the case you are in an asynchronous transfer

Martinhoff-maker commented on drivers/spi/spi_silabs_siwx91x_gspi.c:
```diff
@@ -592,6 +608,45 @@ static int gspi_siwx91x_release(const struct device *dev, const struct spi_confi
+			return ret;
+		}
+
+		ret = pinctrl_apply_state(cfg->pcfg, PINCTRL_STATE_DEFAULT);
+		if (ret < 0 && ret != -ENOENT) {
+			return ret;
+		}
```
Don't you need to recall spi_context_cs_configure_all? or is it not needed with the upcoming PM implementation of GPIO ?

fimohame replied:
addressed

asmellby commented on drivers/spi/spi_silabs_siwx91x_gspi.c:
```diff
@@ -561,6 +571,10 @@ static int gspi_siwx91x_transceive(const struct device *dev, const struct spi_co
 	}
 
+	if (!asynchronous) {
+		pm_device_runtime_put(dev);
```
Unless I'm misunderstanding something, this doesn't seem correct to me. If DMA is used for a blocking transfer, won't the runtime_put already have happened in the DMA complete callback? It seems to me like this should be done inside the `else` above, and not be related to `asynchronous` at all.

fimohame replied:
Your correct. It's been addressed now
